packages/SwingSet/src/controller/controller.js

```diff
@@ -167,340 +168,400 @@ export async function makeSwingsetController(
+   * a `seconds` property valued with the total elapsed duration in seconds.
    *
-   * @param {EndoZipBase64Bundle} bundle
-   * @param {BundleID} [allegedBundleID]
-   * @returns {Promise<BundleID>}
+   * @template T
+   * @template {unknown[]} A
+   * @param {readonly [startLabel: string, endLabel: string]} labels
+   * @param {SlogDurationProps} startProps for both slog entries
+   * @param {(finish: (extraProps?: SlogDurationProps) => void, ...args: A) => (T | Promise<T>)} fn
+   * @param {unknown[] & A} args
+   * @returns {Promise<T>}
    */
-  async function validateAndInstallBundle(bundle, allegedBundleID = undefined) {
-    // TODO The following assertion may be removed when checkBundle subsumes
-    // the responsibility to verify the permanence of a bundle's properties.
-    // https://github.com/endojs/endo/issues/1106
-    mustMatch(bundle, endoZipBase64Sha512Shape);
-    await checkBundle(bundle, computeSha512, allegedBundleID);
-    const { endoZipBase64Sha512 } = bundle;
-    assert.typeof(endoZipBase64Sha512, 'string');
-    const bundleID = `b1-${endoZipBase64Sha512}`;
-    if (allegedBundleID !== undefined) {
-      bundleID === allegedBundleID ||
-        Fail`alleged bundleID ${allegedBundleID} does not match actual ${bundleID}`;
+  const slogDuration = async (labels, startProps, fn, ...args) => {
+    const [startLabel, endLabel] = labels;
+    const props = { ...startProps };
+    if (hasOwn(props, 'type') || hasOwn(props, 'seconds')) {
+      console.error('startProps must not include "type" or "seconds"');
+      delete props.type;
+      delete props.seconds;
     }
-    await kernel.installBundle(bundleID, bundle);
-    return bundleID;
-  }
-
-  // the kernel won't leak our objects into the Vats, we must do
-  // the same in this wrapper
-  const controller = harden({
-    log(str) {
-      kernel.log(str);
-    },
-
-    writeSlogObject,
+    let cause;
+    let finished = false;
+    /** @type {(extraProps?: SlogDurationProps) => void} */
+    const finish = extraProps => {
+      const seconds = (performance.now() - t0) / 1000;
+      if (finished) {
+        // `finish` should only be called once.
+        // Log a stack-bearing error instance, but throw something more opaque.
+        const msg = `slog event ${startLabel} already finished; ignoring props ${q(extraProps || {})}`;
+        const err = Error(msg);
+        console.error(err);
+        Fail`slog event ${startLabel} already finished`;
+      }
+      finished = true;
+      if (extraProps) {
+        // Preserve extraProps as an atomic unit by deleting prior occurrences.
+        for (const name of Object.keys(extraProps)) delete props[name];
+        if (hasOwn(extraProps, 'type') || hasOwn(extraProps, 'seconds')) {
+          console.error(
+            `extraProps ${q(extraProps)} must not include "type" or "seconds"`,
+          );
+          const { type: _, seconds: __, ...validProps } = extraProps;
```

Comment:
This suggestion is up to you to take or leave as you will: drowsy programmers don't easily see the difference between `_`, `__`, `___`, and the conventional disambiguation I've seen is to use `_`, `_2`, `_3`... 

```suggestion
          const { type: _, seconds: _2, ...validProps } = extraProps;
```